File: algorithms/668. Kth Smallest Number in Multiplication Table/solution.py
class Solution(object):
    def findKthNumber(self, m, n, k):
        """
        :type m: int
        :type n: int
        :type k: int
        :rtype: int
        """
        lo, hi = 1, m * n
        while lo < hi:
            mid = lo + (hi - lo) / 2
            if sum(min(mid / i, n) for i in xrange(1, m + 1)) >= k:
                hi = mid
            else:
                lo = mid + 1
        return lo

# 69 / 69 test cases passed.
# Status: Accepted
# Runtime: 1046 ms

File: algorithms/668. Kth Smallest Number in Multiplication Table/solution.cc
class Solution {
public:
    int findKthNumber(int m, int n, int k) {
        int lo = 1, hi = m * n;
        while ( lo < hi ) {
            int mid = lo + (hi - lo) / 2, cnt = 0;
            for ( int i = 1; i <= m; ++i )
                cnt += min(mid / i, n);
            if ( cnt >= k )
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }
};

// 69 / 69 test cases passed.
// Status: Accepted
// Runtime: 26 ms
